src/sentry_core.c

```diff
@@ -448,18 +449,36 @@ sentry__capture_event(sentry_value_t event)
 }
 
 bool
-sentry__should_send_transaction(sentry_value_t tx_cxt)
+sentry__should_send_transaction(
+    sentry_value_t tx_cxt, sentry_sampling_context_t *sampling_ctx)
 {
     sentry_value_t context_setting = sentry_value_get_by_key(tx_cxt, "sampled");
     if (!sentry_value_is_null(context_setting)) {
-        return sentry_value_is_true(context_setting);
+        bool sampled = sentry_value_is_true(context_setting);
+        sampling_ctx->parent_sampled = sentry_malloc(sizeof(bool));
+        *sampling_ctx->parent_sampled = sampled;
     }
 
     bool send = false;
     SENTRY_WITH_OPTIONS (options) {
-        send = sentry__roll_dice(options->traces_sample_rate);
-        // TODO(tracing): Run through traces sampler function if rate is
-        // unavailable.
+        if (options->traces_sampler) {
+            const double result
+                = ((sentry_traces_sampler_function)options->traces_sampler)(
+                    sampling_ctx->transaction_context,
+                    sampling_ctx->custom_sampling_context,
+                    sampling_ctx->parent_sampled);
+            send = sentry__roll_dice(result);
+        } else {
+            if (sampling_ctx->parent_sampled != NULL) {
+                send = *sampling_ctx->parent_sampled;
+            } else {
+                send = sentry__roll_dice(options->traces_sample_rate);
+            }
+        }
+    }
+    if (sampling_ctx->parent_sampled != NULL) {
+        sentry_free(sampling_ctx->parent_sampled);
+        sampling_ctx->parent_sampled = NULL;
```

Comment:
You can eliminate the `free` by pointing to the stack variable. Writing `NULL` at the end is still excellent citizenship.